File: python/66.加一.py
#
# @lc app=leetcode.cn id=66 lang=python3
#
# [66] 加一
#
from typing import List


# @lc code=start
class Solution:
    # 算9
    def plusOne(self, digits: List[int]) -> List[int]:
        n = len(digits)
        for i in range(n - 1, -1, -1):
            if digits[i] != 9:
                digits[i] += 1
                for j in range(i + 1, n):
                    digits[j] = 0
                return digits
        return [1] + [0] * n

    # 取模运算
    def plusOne1(self, digits: List[int]) -> List[int]:
        index = len(digits) - 1
        digits[index] += 1
        carry = 0
        while index >= 0:
            total = (digits[index] + carry)
            digits[index] = total % 10
            carry = total // 10
            index -= 1
        if carry > 0:
            ret = [0] * (len(digits) + 1)
            ret[0] = carry
            ret[1:] = digits
            return ret
        return digits


# @lc code=end

File: cpp/66.加一.cpp
/*
 * @lc app=leetcode.cn id=66 lang=cpp
 *
 * [66] 加一
 */
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
// @lc code=start
class Solution
{
public:
    vector<int> plusOne(vector<int>& digits)
    {
        return method2(digits);
    }

    // 利用只+1的性质，所以进位也只可能是1
    vector<int> method2(vector<int>& digits) {
        for (int i = digits.size() - 1; i >= 0; i--) {
            digits[i] = (digits[i] + 1) % 10;
            if (digits[i] != 0) {
                return digits;
            }
        }
        // 能走到这里，说明每一位都进位了(e.g. 999)，结果的长度会+1
        vector<int> ans(digits.size() + 1, 0);
        ans[0] = 1;
        return ans;
    }

    // 相对通用的解法，加多少都可以
    vector<int> method1(vector<int>& digits) {
        vector<int> ans;
        int carry = 1;
        for (int i = digits.size() - 1; i >= 0; i--) {
            int cur = (digits[i] + carry) % 10;
            carry   = (digits[i] + carry) / 10;
            ans.push_back(cur);
        }
        if (carry > 0) {
            ans.push_back(carry);
        }

        reverse(ans.begin(), ans.end());
        return ans;
    }
};
// @lc code=end
